kmcd-prelude.md

```diff
@@ -275,5 +276,269 @@ module KMCD-GEN
 
     rule <k> GSS | GSS' => #if R modInt 2 ==K 0 #then GSS #else GSS' #fi ... </k>
          <random> R => randInt(R) </random>
+
+    syntax GenStep ::= GenTimeStep
+    syntax GenTimeStep ::= "GenTimeStep"
+ // ------------------------------------
+    rule <k> GenTimeStep => LogGen ( TimeStep ((I modInt 2) +Int 1) ) ... </k>
+         <random> I => randInt(I) </random>
+
+    syntax GenStep ::= GenVatStep
+    syntax GenVatStep ::= "GenVatFrob"
+                        | "GenVatFrob" CDPID
+                        | "GenVatFrob" CDPID Wad [prefer]
+                        | "GenVatMove"
+                        | "GenVatMove" Address
+                        | "GenVatMove" Address Address
+                        | "GenVatHope"
+                        | "GenVatHope" Address
+                        | "GenVatHope" Address Address
+ // --------------------------------------------------
+    rule <k> GenVatFrob => GenVatFrob chooseCDPID(I, keys_list(VAT_URNS)) ... </k>
+         <random> I => randInt(I) </random>
+         <vat-urns> VAT_URNS </vat-urns>
+      requires size(VAT_URNS) >Int 0
+
+    rule <k> GenVatFrob CDPID => GenVatFrob CDPID ((2 *Rat randRatBounded(I, VAT_GEM)) -Rat VAT_GEM) ... </k>
+         <random> I => randInt(I) </random>
+         <vat-gem> ... CDPID |-> VAT_GEM ... </vat-gem>
+
+    rule <k> GenVatFrob { ILKID , ADDRESS } DINK
+          => #fun( DARTBOUND
+                => LogGen ( transact ADDRESS Vat . frob ILKID ADDRESS ADDRESS ADDRESS DINK ((2 *Rat randRatBounded(I, DARTBOUND)) -Rat DARTBOUND) )
+                 ) (((SPOT *Rat (URNINK +Rat DINK)) /Rat RATE) -Rat URNART)
+         ...
+         </k>
+         <random> I => randInt(I) </random>
+         <vat-ilks>
+           ...
+           ILKID |-> Ilk ( ... rate: RATE, spot: SPOT )
+           ...
+         </vat-ilks>
+         <vat-urns>
+           ...
+           { ILKID , ADDRESS } |-> Urn ( ... ink: URNINK, art: URNART )
+           ...
+         </vat-urns>
```

Comment:
GenVatFrob may (in principle) fail here as well if the needed vat-ilks or vat-urns are not found. An additional `owise` rule would do.